nsqconsumer/consumer.go

```diff
@@ -167,84 +168,101 @@ func (c *nsqConsumer) Start(ctx context.Context) func() {
 
 	consumer.SetLogger(log.New(os.Stderr, fmt.Sprintf("[nsq-consumer(%s)]", c.Topic), log.Flags()), nsq.LogLevelWarning)
 
-	consumer.AddConcurrentHandlers(nsq.HandlerFunc(func(message *nsq.Message) (err error) {
-		defer func() {
-			if r := recover(); r != nil {
-				var errRecovered error
-				switch value := errRecovered.(type) {
-				case error:
-					errRecovered = value
-				default:
-					errRecovered = errgo.Newf("%v", value)
-				}
-				err = errgo.Newf("recover panic from nsq consumer: %+v", errRecovered)
-				c.logger.WithError(errRecovered).WithFields(logrus.Fields{"stacktrace": string(debug.Stack())}).Error("recover panic")
-			}
-		}()
+	consumer.AddConcurrentHandlers(nsq.HandlerFunc(c.nsqHandler), c.MaxInFlight)
 
-		if len(message.Body) == 0 {
-			err := errgo.New("body is blank, re-enqueued message")
-			c.logger.WithError(err).Error("blank message")
-			return err
-		}
-		var msg NsqMessageDeserialize
-		err = json.Unmarshal(message.Body, &msg)
-		if err != nil {
-			c.logger.WithError(err).Error("failed to unmarshal message")
-			return err
-		}
-		msg.NsqMsg = message
-
-		msgLogger := c.logger.WithFields(logrus.Fields{
-			"message_id":   fmt.Sprintf("%s", message.ID),
-			"message_type": msg.Type,
-			"request_id":   msg.RequestID,
-		})
-
-		ctx := context.WithValue(
-			context.WithValue(
-				context.Background(), "request_id", msg.RequestID,
-			), "logger", msgLogger,
-		)
-
-		if msg.At != 0 {
-			now := time.Now().Unix()
-			delay := msg.At - now
-			if delay > 0 {
-				return c.postponeMessage(ctx, msgLogger, msg, delay)
+	if err = consumer.ConnectToNSQLookupds(c.NsqLookupdURLs); err != nil {
+		c.logger.WithError(err).Fatalf("Fail to connect to NSQ lookupd")
+	}
+
+	return func() {
+		consumer.Stop()
+		// block until stop process is complete
+		<-consumer.StopChan
+	}
+}
+
+func (c *nsqConsumer) nsqHandler(message *nsq.Message) (err error) {
+	defer func() {
+		if r := recover(); r != nil {
+			var errRecovered error
+			switch value := errRecovered.(type) {
+			case error:
+				errRecovered = value
+			default:
+				errRecovered = errgo.Newf("%v", value)
 			}
+			err = errgo.Newf("recover panic from nsq consumer: %+v", errRecovered)
+			c.logger.WithError(errRecovered).WithFields(logrus.Fields{"stacktrace": string(debug.Stack())}).Error("recover panic")
 		}
+	}()
+
+	if len(message.Body) == 0 {
+		err := errgo.New("body is blank, re-enqueued message")
+		c.logger.WithError(err).Error("blank message")
+		return err
+	}
+	var msg NsqMessageDeserialize
+	err = json.Unmarshal(message.Body, &msg)
+	if err != nil {
+		c.logger.WithError(err).Error("failed to unmarshal message")
```

Comment:
picky picky: while you are on it :)
```suggestion
		c.logger.WithError(err).Error("Fail to unmarshal message")
```